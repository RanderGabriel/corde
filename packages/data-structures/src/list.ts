/**
 * List if an encapsulation of [Array](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Array).
 * It's contains some others functions that are no present in Array itself
 * While provide access direct to main Array functions.
 *
 * The main goal is to reduce the acess to no relevant functions provided by Array and
 * specialize this class in iteration adition, reduction and search of elements.
 */
export class List<T> {
  private _data: T[];

  /**
   * Initialize a new List.
   */
  constructor() {
    this._data = [];
  }

  /**
   * Converts a array to List.
   * It do not modify the current values of the list.
   * @param data Array with values to be converted.
   * @returns A new instance of List. (The list has no references this instance).
   */
  public arrayToList(data: T[]): List<T> {
    const newList = new List<T>();
    for (const val of data) {
      newList.add(val);
    }
    return newList;
  }

  /**
   * Get the first elements that match with the predicate
   * does the same of [Array.find](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Array/find).
   * @param predicate Comparator to find the element.
   * @returns The first element of the list that match with the predicate,
   * or null.
   */
  public single(predicate: (value: T) => boolean): T {
    return this._data.find(predicate) ?? null;
  }

  /**
   * Same of
   * [Array.filter](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Array/filtro).
   * Finds a List of elements that match with a predicate.
   * @param predicate Comparator to find the elements.
   * @returns An list of all elements that match with the predicate.
   * returns an empty list if no element was found.
   */
  public filter(predicate: (value: T) => boolean): List<T> {
    const arr = this._data.filter(predicate);
    return this.arrayToList(arr);
  }

  /**
   * Picks the **first** element of the list, or null if no element exists.
   * @returns The first element or undefined if no element exists.
   */
  public first(): T | undefined {
    return this._data[0];
  }

  /**
   * Picks the **last** element of the list, or undefined if no element exists.
   * @returns The last element of the list of undefined if no element exists.
   */
  public last(): T | undefined {
    return this._data[this._data.length - 1];
  }

  /**
   * Removes all elements of the list.
   * @returns This instance.
   */
  public clear(): List<T> {
    this._data = [];
    return this;
  }

  /**
   * Copy all elements of the list to a new List.
   * @returns The cloned list.
   */
  public clone(): List<T> {
    return this.arrayToList(this._data.map((d) => d));
  }

  /**
   * Do a operation for each element of the list.
   * Same of [Array.foreach](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach).
   * @param operation Operation to execute in each element.
   * @returns this instance.
   */
  public each(operation: (val: T) => void): List<T> {
    for (const obj of this._data) {
      operation(obj);
    }
    return this;
  }

  /**
   * Select all properties informed in parameter that are contained in
   * the List object.
   * @param keys Properties of T to be picked.
   * @returns All picked properties of T in a new List.
   * Returns an empty list of no element was found.
   */
  public select<K extends keyof T>(...keys: K[]): List<Pick<T, K>> {
    const newList = new List<Pick<T, K>>();
    for (const obj of this._data) {
      const copy = {} as Pick<T, K>;
      keys.forEach((key) => (copy[key] = obj[key]));
      newList.add(copy);
    }
    return newList;
  }

  /**
   * Add a new element to list.
   * Note: undefined or null elements are not added.
   * @param values The new element.
   * @returns This instance.
   */
  public add(...values: T[]): List<T> {
    for (const val of values) {
      if (val) {
        this._data.push(val);
      }
    }
    return this;
  }

  /**
   * Checks if the list has at least one element.
   * @returns True if there are elements in list, false if don't.
   */
  public any() {
    return this._data.length > 0;
  }

  /**
   * Removes a element from the list.
   * @param data Element to be removed.
   * @returns This instance.
   */
  public remove(data: T): List<T>;
  /**
   * Removes an array of elements from the list.
   * @param data Array of elements.
   * @returns This instance.
   */
  public remove(data: T[]): List<T>;
  /**
   * Removes an list of elements from the list.
   * @param data List of elements.
   * @returns This instance.
   */
  public remove(data: List<T>): List<T>;
  public remove(data: T | T[] | List<T>) {
    if (!data) {
      return this;
    }
    if (this._isArray(data) || data instanceof List) {
      (data as T[]).forEach((d) => this._removeItem(d));
      return this;
    } else {
      this._removeItem(data);
      return this;
    }
  }

  /**
   * Get a element based on its **index**.
   * @param index Position in list of the element
   * @returns The element in the position informed.
   */
  public get(index: number): T;
  /**
   * Get a list of elements based in their **index**.
   * @param index Position of each element in list.
   * @returns A list with each element informed in array.
   * No existing elements will be ignored in adition on list.
   */
  public get(...index: number[]): List<T>;
  public get(index: number | number[]): T | List<T> {
    if (this._isArray(index)) {
      const values = new List<T>();
      for (const i of index) {
        const element = this._data[i];
        if (element) {
          values.add(element);
        }
      }
      return values;
    } else {
      return this._data[index];
    }
  }

  /**
   * Returns the index of the first element that matchs with the informed element.
   * Same of [Array.indexOf](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Array/indexOf).
   * @param element Element to have its index searched.
   * @returns Index of the searched element.
   */
  public indexOf(element: T): number {
    return this._data.indexOf(element);
  }

  public has(element: T) {
    return !!this._data.find((d) => d === element);
  }

  private _removeItem(data: T) {
    const index = this._data.indexOf(data, 0);
    if (index > -1) {
      this._data.slice(index, 1);
    }
  }

  private _isArray(content: any | any[]): content is any[] {
    return Array.isArray(content);
  }
}
