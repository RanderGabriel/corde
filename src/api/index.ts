export * from "./afterAll";
export * from "./beforeStart";
export * from "./expect";
export * from "./expectMatches";
export * from "./group";
export * from "./matcher";
export * from "./interfaces";
export * from "./test";
export * from "./afterEach";
export * from "./beforeEach";
export * from "./sendMessage";
export * from "./getRole";
export * from "./createRole";
