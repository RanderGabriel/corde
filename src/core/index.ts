import { CordeBot } from "./cordeBot";

export { CordeBot } from "./cordeBot";
