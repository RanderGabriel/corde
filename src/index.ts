/**
 * @file Automatically generated by barrelsby.
 */

 export * from "./cli";
export * from "./erros/index";
