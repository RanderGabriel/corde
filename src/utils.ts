import { ParameterNotFoundError } from './errors';

export function clone(object: object | object[]): object | object[] {
  if (!object) {
    throw new ParameterNotFoundError('Object can not be undefined');
  }

  if (object as object[]) {
    const copyArray = [];
    for (const i in object as []) {
      copyArray.push(copyObject((object as [])[i]));
    }
    return copyArray;
  } else {
    return copyObject(object);
  }
}

function copyObject(object: object) {
  return JSON.parse(JSON.stringify(object));
}
