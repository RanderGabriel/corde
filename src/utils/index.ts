export * from "./guid";
export * from "./colors";
export * from "./permission";
export * from "./executeWithTimeout";
export * from "./isFunction";
export * from "./bitfield";
export * from "./isPrimitiveValue";
export * from "./wait";
export * from "./pick";
export * from "./tryImport";
export * from "./formatObject";
export * from "./diff";
export * from "./isPromise";
export * from "./dateDiff";
export * from "./logUpdate";
export * from "./stringIsNullOrEmpty";
export * from "./timer";
