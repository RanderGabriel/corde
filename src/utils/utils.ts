import chalk from 'chalk';
import { ParameterNotFoundError } from '../errors';

/**
 * Clone a object using JSON.parse
 *
 * @param object Object to be cloned
 */
export function clone<T>(object: T | T[]): T | T[] {
  if (!object) {
    throw new ParameterNotFoundError('Object can not be undefined');
  }
  return JSON.parse(JSON.stringify(object));
}

/**
 * Pick some properties of a object
 *
 * @see https://www.typescriptlang.org/docs/handbook/utility-types.html#picktk
 * @param obj Object to get its properties
 * @param keys Properties that must be got
 */
export function pick<T, K extends keyof T>(obj: T, ...keys: K[]): Pick<T, K> {
  const copy = {} as Pick<T, K>;
  keys.forEach((key) => (copy[key] = obj[key]));
  return copy;
}

/**
 * Finish the process with code 1
 * displaying a default message in case of no data in parameter.
 *
 * @param msg Message to be displayed before exit the process
 */
export function exitProcessWithError(msg?: string) {
  if (msg) {
    console.log(`\n${chalk.red('error')} ${msg}`);
  } else {
    console.log(`\n${chalk.red('error')} Command failed with exit code 1`);
  }
  process.exit(1);
}
