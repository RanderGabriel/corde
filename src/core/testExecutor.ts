import chalk from "chalk";
import { runtime } from "../common/runtime";
import { testCollector } from "../common/testCollector";
import { printHookErrors } from "../common/printHookError";
import {
  MESSAGE_TAB_SPACE,
  TAG_FAIL,
  TAG_PASS,
  TAG_PENDING,
  TEST_FAIL_ICON,
  TEST_PASSED_ICON,
  TEST_RUNNING_ICON,
  TEXT_EMPTY,
  TEXT_FAIL,
  TEXT_PASS,
} from "../consts";
import { Queue } from "../data-structures";
import {
  Group,
  RunnerReport,
  SemiRunnerReport,
  Test,
  TestFile,
  TestReport,
  VoidLikeFunction,
} from "../types";
import { formatObject, stringIsNullOrEmpty, Timer } from "../utils";
import { LogUpdate } from "../utils";

export class TestExecutor {
  private _logUpdate: LogUpdate;

  constructor(logUpdate: LogUpdate) {
    this._logUpdate = logUpdate;
  }

  async runTestsAndPrint(testFiles: TestFile[]): Promise<RunnerReport> {
    const testsTimer = new Timer();
    testsTimer.start();

    const semiReport: SemiRunnerReport = {
      totalTests: 0,
      totalTestFiles: 0,
      totalTestsPassed: 0,
      totalTestsFailed: 0,
      totalTestFilesPassed: 0,
      totalTestFilesFailed: 0,
      totalEmptyTestFiles: 0,
      totalEmptyTests: 0,
    };

    for (const testFile of testFiles) {
      if (testFile.isEmpty) {
        this._logUpdate.append(`${TAG_PENDING("EMPTY")}  ${testFile.path}`);
        this._logUpdate.persist();
        semiReport.totalEmptyTestFiles++;
        semiReport.totalTestFiles++;
      } else {
        if (await this.executeTestFile(testFile, semiReport)) {
          semiReport.totalTestFilesPassed++;
        } else {
          semiReport.totalTestFilesFailed++;
        }
      }
    }
    const testsDiff = testsTimer.stop();

    return {
      testTimer: testsDiff[0],
      ...semiReport,
    };
  }

  private async executeTestFile(testFile: TestFile, semiReport: SemiRunnerReport) {
    const testFileTimer = new Timer();
    semiReport.totalTestFiles++;
    let fileHasPassed = true;

    testFileTimer.start();
    const logIndex = this._logUpdate.append(`${TAG_PENDING()}  ${testFile.path}`);

    for (const group of testFile.groups) {
      if (!(await this.executeGroup(group, semiReport))) {
        fileHasPassed = false;
      }
    }

    const _diff = testFileTimer.stop();

    let fileLabel = TAG_PASS();
    let fileNameLabel = testFile.path;

    if (!fileHasPassed) {
      fileLabel = TAG_FAIL();
      fileNameLabel = chalk.red(fileNameLabel);
    }

    this._logUpdate.updateLine(
      logIndex,
      `${fileLabel}  ${fileNameLabel}   ${chalk.cyan(_diff[0])}`,
    );

    this._logUpdate.persist();

    return fileHasPassed;
  }

  private async executeGroup(group: Group, semiReport: SemiRunnerReport) {
    const tests = this.getAssertionPropsFromGroup(group);
    let fileHasPassed = true;
    for (const test of tests) {
      if (!(await this.executeTest(test, semiReport, group.name))) {
        fileHasPassed = false;
      }
    }
    return fileHasPassed;
  }

  private async executeTest(
    test: Test,
    semiReport: SemiRunnerReport,
    groupName?: string | number | boolean,
  ) {
    const testTimer = new Timer();
    let fileHasPassed = true;
    testTimer.start();
    let logPosition = 0;

    const testText = this.createTestText(test.name);
    logPosition = this._logUpdate.appendLine(testText);

    const reports = await this.runTest(test);

    if (reports.length === 0) {
      semiReport.totalEmptyTests++;
      semiReport.totalTests++;
    }

    const testDiff = testTimer.stop();

    const testNameLabel = this.testReportLabelFunction(reports);

    const formatedGroupName = !stringIsNullOrEmpty(groupName) ? groupName + " -> " : "";

    if (stringIsNullOrEmpty(test.name)) {
      this._logUpdate.updateLine(
        logPosition,
        `${MESSAGE_TAB_SPACE}${testNameLabel(
          formatedGroupName + "<empty test name>",
        )}   ${chalk.cyan(testDiff[0])}`,
      );
    } else {
      this._logUpdate.updateLine(
        logPosition,
        `${MESSAGE_TAB_SPACE}${testNameLabel(formatedGroupName + test.name)}   ${chalk.cyan(
          testDiff[0],
        )}`,
      );
    }

    if (!this.printTestsReportAndUpdateRunnerReport(reports, semiReport)) {
      fileHasPassed = false;
    }

    test.testsReports = reports;
    return fileHasPassed;
  }

  private printTestsReportAndUpdateRunnerReport(
    reports: TestReport[],
    semiReport: SemiRunnerReport,
  ) {
    let fileHasPassed = true;
    for (const report of reports) {
      semiReport.totalTests++;
      if (!this.printReportAndUpdateRunnerReport(report, semiReport)) {
        fileHasPassed = false;
      }
    }
    return fileHasPassed;
  }

  private testReportLabelFunction(reports: TestReport[]) {
    if (reports.length === 0) {
      return (text: string) => TEXT_EMPTY(" " + TEST_RUNNING_ICON + " " + text + " (empty)");
    }

    if (reports.some((report) => !report.pass)) {
      return (text: string) => TEXT_FAIL(TEST_FAIL_ICON + " " + text);
    }
    return (text: string) => TEXT_PASS(TEST_PASSED_ICON + " " + text);
  }

  private printReportAndUpdateRunnerReport(report: TestReport, semiReport: SemiRunnerReport) {
    if (report.pass) {
      semiReport.totalTestsPassed++;
      return true;
    }

    semiReport.totalTestsFailed++;
    this.printReportData(report);

    return false;
  }

  private printReportData(report: TestReport) {
    if (report.message) {
      this._logUpdate.appendLine(report.message);
    }

    if (!report.pass && report.trace) {
      this._logUpdate.appendLine(report.trace);
    }
  }

  private createTestText(testName: string | number | boolean) {
    if (stringIsNullOrEmpty(testName)) {
      return this.createTestTextByStatus("<empty test name>");
    }
    return this.createTestTextByStatus(testName);
  }

  private createTestTextByStatus(testName: string | number | boolean) {
    const icon = TEST_RUNNING_ICON;
    return `${MESSAGE_TAB_SPACE}${icon} ${testName}`;
  }

  async runTest(test: Test) {
    const reports: TestReport[] = [];

    // before e after hooks will run just one time
    // for test structure (if the hook fail)
    let keepRunningBeforeEachFunctions = true;
    let keepRunningAfterEachFunctions = true;

    for (const testfn of test.testsFunctions) {
      keepRunningBeforeEachFunctions = await this.executeHookFunctionIfPossible(
        keepRunningBeforeEachFunctions,
        testCollector.beforeEachFunctions,
      );

      let _report: TestReport;
      try {
        _report = await runtime.injectBot(testfn);
      } catch (error) {
        _report = {
          pass: false,
          message: this.getErrorMessage(error),
        };
      }

      keepRunningAfterEachFunctions = await this.executeHookFunctionIfPossible(
        keepRunningAfterEachFunctions,
        testCollector.afterEachFunctions,
      );

      reports.push(_report);
    }
    return reports;
  }

  private async executeHookFunctionIfPossible(
    keepRunning: boolean,
    queues: Queue<VoidLikeFunction>,
  ) {
    if (keepRunning) {
      const _functionErrors = await queues.executeWithCatchCollectAsync();
      if (_functionErrors && _functionErrors.length) {
        printHookErrors(_functionErrors);
        return false;
      }
      return true;
    }
    return false;
  }

  private getErrorMessage(error: any) {
    if (error instanceof Error) {
      return error.message;
    }

    return formatObject(error);
  }

  private getAssertionPropsFromGroup(group: Group) {
    const assertions: Test[] = [];
    if (group.tests) {
      group.tests.forEach((test) => {
        assertions.push(...this.getAssertionsPropsFromTest(test));
      });
    }

    if (group.subGroups) {
      group.subGroups.forEach((subGroup) => {
        const subGroupTests = this.getAssertionPropsFromGroup(subGroup);
        assertions.push(...subGroupTests);
      });
    }

    return assertions;
  }

  private getAssertionsPropsFromTest(test: Test) {
    const tests: Test[] = [test];

    if (test.subTests) {
      test.subTests.forEach((subtest) => {
        tests.push(...this.getAssertionsPropsFromTest(subtest));
      });
    }

    return tests;
  }
}
