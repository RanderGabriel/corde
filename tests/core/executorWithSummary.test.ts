import { summary } from "../../src/core";
import { TestExecutor } from "../../src/core/testExecutor";
import { buildReportMessage, LogUpdate } from "../../src/utils";
import { mockTimer } from "../mocks/mockTimer";
import { generateTestFile, removeANSIColorStyle, TestFileGeneratorInfo } from "../testHelper";

let logUpdate: LogUpdate;
let testRunner: TestExecutor;

mockTimer();

beforeEach(() => {
  logUpdate = new LogUpdate();
  testRunner = new TestExecutor(logUpdate);
});

it("should print report for 1 test file, 1 test clausure and 1 failed function and 1 passed function", async () => {
  const data: TestFileGeneratorInfo = {
    amountOfTestFiles: 1,
    testFunctionsReport: [
      {
        pass: false,
        message: buildReportMessage("expected: hi", "received: hi!"),
      },
      { pass: true },
    ],
    amountOfTests: 1,
  };

  const tests = generateTestFile(data);

  const report = await testRunner.runTestsAndPrint(tests);
  const summaryStder = summary.print(report);
  expect(removeANSIColorStyle(logUpdate.stder + summaryStder)).toMatchSnapshot();
});
