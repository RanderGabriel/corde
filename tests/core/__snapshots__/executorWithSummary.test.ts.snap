// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should print report for 1 test file, 1 test clausure and 1 failed function and 1 passed function 1`] = `
"
 FAIL   /tests/file1.test.ts   10ms
   ❌ group -> test case1   3ms

   expected: hi
   received: hi!

Test Files: 1 failed. 1 total
Tests:      1 passed and 1 failed. 2 total
Time:       14ms"
`;
